### ISSUE TITLE
[3.1] Query error when using HABTM relationship through custom pivot table

### ISSUE THREAD

**Issue by sebio** [cakephp/cakephp#7549]
Let's say I have two models, `Articles` and `Tags`, with a belongsToMany relationship between them through a custom join model `ArticlesTags`. Here's my setup:

```php
class ArticlesTable extends Table
{
    public function initialize(array $config)
    {
        $this->table('cake_articles');
        $this->belongsToMany('Tags', [
            'through'          => 'ArticlesTags',
            'foreignKey'       => 'article_id',
            'targetForeignKey' => 'tag_id'
        ]);
    }
}
```

```php
class TagsTable extends Table
{
    public function initialize(array $config)
    {
        $this->table('cake_tags');
        $this->belongsToMany('Articles', [
            'through'          => 'ArticlesTags',
            'foreignKey'       => 'tag_id',
            'targetForeignKey' => 'article_id'
        ]);
    }
}
```

```php
class ArticlesTagsTable extends Table
{
    public function initialize(array $config)
    {
        $this->table('cake_articles_tags');
        $this->belongsTo('Articles');
        $this->belongsTo('Tags');
    }
}
```

When I execute the next two find instructions, the first one builds the query correctly and returns the expected result, but the second one throws and exception:

```php
$articles = TableRegistry::get('Articles')->find()->contain('Tags')->all();
$tags     = TableRegistry::get('Tags')->find()->contain('Articles')->all();
```

```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'ArticlesTags.cake_article_id' in 'field list'
```

If I comment the first find instruction, then the second works just as expected. Here's the generated SQL from DebugKit:

```sql
SELECT 
  Articles.id AS `Articles__id`, 
  Articles.title AS `Articles__title` 
FROM 
  cake_articles Articles

SELECT 
  ArticlesTags.tag_id AS `ArticlesTags__tag_id`, 
  ArticlesTags.id AS `ArticlesTags__id`, 
  ArticlesTags.article_id AS `ArticlesTags__article_id`, 
  Tags.id AS `Tags__id`, 
  Tags.name AS `Tags__name` 
FROM 
  cake_tags Tags 
  INNER JOIN cake_articles_tags ArticlesTags ON Tags.id = (ArticlesTags.tag_id) 
WHERE 
  ArticlesTags.article_id in ('1', '2')

SELECT 
  Tags.id AS `Tags__id`, 
  Tags.name AS `Tags__name` 
FROM 
  cake_tags Tags

SELECT 
  ArticlesTags.cake_article_id AS `ArticlesTags__cake_article_id`, 
  ArticlesTags.id AS `ArticlesTags__id`, 
  ArticlesTags.article_id AS `ArticlesTags__article_id`, 
  ArticlesTags.tag_id AS `ArticlesTags__tag_id`, 
  Articles.id AS `Articles__id`, 
  Articles.title AS `Articles__title` 
FROM 
  cake_articles Articles 
  INNER JOIN cake_articles_tags ArticlesTags ON Articles.id = (ArticlesTags.cake_article_id) 
WHERE 
  ArticlesTags.tag_id in ('1', '2', '3')
```

I seems that behind the scene, the ORM omits entirely the foreign key specified in the associations setup and sets the field based on table name (not even model name, wich should be the table alias, as documented).

**Comment by lorenzo** [comment]
The problem are the `belongsTo()` calls in the `ArticlesTabsTable` class. You also need to specify the `foreignKey` option when defining the association

**Comment by sebio** [comment]
@lorenzo I also tried exactly as you suggested using the following calls for belongsTo association but without any success.
```php
        $this->belongsTo('Articles', ['foreignKey' => 'article_id']);
        $this->belongsTo('Tags', ['foreignKey' => 'tag_id']);
```
I found that this problem started with 3.0.4 after a fix for another issue described here https://github.com/cakephp/cakephp/pull/6344, and it has been around since then.

In the meantime I also found that internally `\ORM\Association\BelongsToMany::junction()` creates an Articles hasMany ArticlesTags association to Articles table when calling the first find. If I remove this assoc before the second find, it gets recreated but with the proper foreignKey this time.
```php
$articles = TableRegistry::get('Articles')->find()->contain('Tags')->all();
TableRegistry::get('Articles')->associations()->remove('ArticlesTags');
$tags = TableRegistry::get('Tags')->find()->contain('Articles')->all();
```
This is a simple workaround I can use to continue my work until a proper fix. Thanks!
